src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -162,72 +124,105 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
+        private static bool IsAccessedBeforeAssignment(
+            SemanticModel semanticModel,
+            ISymbol localVariable,
+            StatementSyntax declarationStatement,
+            StatementSyntax usageStatement,
+            SyntaxList<StatementSyntax> blockStatements,
             CancellationToken cancellationToken)
         {
-            var semanticModel = syntaxContext.SemanticModel;
-            var localVariable = semanticModel.GetDeclaredSymbol(declarator);
-
             var isAssigned = false;
             var isAccessedBeforeAssignment = false;
 
-            CheckDefiniteAssignment(
-                semanticModel, localVariable, ifStatement.Else,
-                out isAssigned, out isAccessedBeforeAssignment,
-                cancellationToken);
+            var usageIndex = blockStatements.IndexOf(usageStatement);
+            var declarationIndex = blockStatements.IndexOf(declarationStatement);
 
-            if (isAccessedBeforeAssignment)
+            // since we're going to remove this declaration-statement, 
+            // we need to first ensure that it's not used up to the target statement.
+            for (var index = declarationIndex + 1; index < usageIndex; index++)
             {
-                return true;
+                CheckDefiniteAssignment(
+                    semanticModel, localVariable, blockStatements[index],
+                    out isAssigned, out isAccessedBeforeAssignment,
+                    cancellationToken);
+
+                if (isAssigned || isAccessedBeforeAssignment)
+                {
+                    return true;
+                }
             }
 
-            var parentBlock = (BlockSyntax)ifStatement.Parent;
-            var ifIndex = parentBlock.Statements.IndexOf(ifStatement);
-            for (int i = ifIndex + 1, n = parentBlock.Statements.Count; i < n; i++)
+            // in case of an if-statement, we need to check if the variable
+            // is being accessed before assignemnt in the else clause.
+            if (usageStatement is IfStatementSyntax ifStatement)
             {
-                if (!isAssigned)
+                CheckDefiniteAssignment(
+                    semanticModel, localVariable, ifStatement.Else,
+                    out isAssigned, out isAccessedBeforeAssignment,
+                    cancellationToken);
+
+                if (isAccessedBeforeAssignment)
                 {
-                    CheckDefiniteAssignment(
-                        semanticModel, localVariable, parentBlock.Statements[i],
-                        out isAssigned, out isAccessedBeforeAssignment,
-                        cancellationToken);
+                    return true;
+                }
 
-                    if (isAccessedBeforeAssignment)
-                    {
-                        return true;
-                    }
+                if (isAssigned)
+                {
+                    return false;
```

Comment:
Nevertheless, I think CheckDefnitiveAssignment is oversimplified. what if the statement is branched, etc? Could we use control flow analysis instead?